pkg/executor/join/semi_join_probe.go

```diff
@@ -0,0 +1,332 @@
+	s.rowIter = nil
+	s.baseJoinProbe.ResetProbe()
+}
+
+func (s *semiJoinProbe) Probe(joinResult *hashjoinWorkerResult, sqlKiller *sqlkiller.SQLKiller) (ok bool, _ *hashjoinWorkerResult) {
+	if joinResult.chk.IsFull() {
+		return true, joinResult
+	}
+
+	joinedChk, remainCap, err := s.prepareForProbe(joinResult.chk)
+	if err != nil {
+		joinResult.err = err
+		return false, joinResult
+	}
+
+	isInCompleteChunk := joinedChk.IsInCompleteChunk()
+	// in case that virtual rows is not maintained correctly
+	joinedChk.SetNumVirtualRows(joinedChk.NumRows())
+	// always set in complete chunk during probe
+	joinedChk.SetInCompleteChunk(true)
+	defer joinedChk.SetInCompleteChunk(isInCompleteChunk)
+
+	hasOtherCondition := s.ctx.hasOtherCondition()
+	if s.isLeftSideBuild {
+		if hasOtherCondition {
+			err = s.probeForLeftSideBuildHasOtherCondition(joinedChk, sqlKiller)
+		} else {
+			err = s.probeForLeftSideBuildNoOtherCondition(sqlKiller)
+		}
+	} else {
+		if hasOtherCondition {
+			err = s.probeForRightSideBuildHasOtherCondition(joinResult.chk, joinedChk, remainCap, sqlKiller)
+		} else {
+			err = s.probeForRightSideBuildNoOtherCondition(joinResult.chk, remainCap, sqlKiller)
+		}
+	}
+	if err != nil {
+		joinResult.err = err
+		return false, joinResult
+	}
+	return true, joinResult
+}
+
+func (s *semiJoinProbe) setIsMatchedRows() {
+	for i, res := range s.selected {
+		if !res {
+			continue
+		}
+
+		s.isMatchedRows[s.rowIndexInfos[i].probeRowIndex] = true
+	}
+}
+
+func (s *semiJoinProbe) probeForLeftSideBuildHasOtherCondition(joinedChk *chunk.Chunk, sqlKiller *sqlkiller.SQLKiller) (err error) {
+	err = s.concatenateProbeAndBuildRows(joinedChk, sqlKiller, false)
+	if err != nil {
+		return err
+	}
+
+	if s.unFinishedProbeRowIdxQueue.IsEmpty() {
+		// To avoid `Previous chunk is not probed yet` error
+		s.currentProbeRow = s.chunkRows
+	}
+
+	meta := s.ctx.hashTableMeta
+	if joinedChk.NumRows() > 0 {
+		s.selected, err = expression.VectorizedFilter(s.ctx.SessCtx.GetExprCtx().GetEvalCtx(), s.ctx.SessCtx.GetSessionVars().EnableVectorizedExpression, s.ctx.OtherCondition, chunk.NewIterator4Chunk(joinedChk), s.selected)
+		if err != nil {
+			return err
+		}
+
+		for index, result := range s.selected {
+			if result {
+				meta.setUsedFlag(*(*unsafe.Pointer)(unsafe.Pointer(&s.rowIndexInfos[index].buildRowStart)))
+			}
+		}
+	}
+
+	return
+}
+
+func (s *semiJoinProbe) probeForLeftSideBuildNoOtherCondition(sqlKiller *sqlkiller.SQLKiller) (err error) {
+	meta := s.ctx.hashTableMeta
+	tagHelper := s.ctx.hashTableContext.tagHelper
+
+	loopCnt := 0
+
+	for s.currentProbeRow < s.chunkRows {
+		if s.matchedRowsHeaders[s.currentProbeRow] != 0 {
+			candidateRow := tagHelper.toUnsafePointer(s.matchedRowsHeaders[s.currentProbeRow])
+			if !meta.isCurrentRowUsedWithAtomic(candidateRow) && isKeyMatched(meta.keyMode, s.serializedKeys[s.currentProbeRow], candidateRow, meta) {
+				meta.setUsedFlag(candidateRow)
+			} else {
+				s.probeCollision++
```

Comment:
should not increase the probeCollision if `meta.isCurrentRowUsedWithAtomic(candidateRow)` is true

Reply:
> should not increase the probeCollision if `meta.isCurrentRowUsedWithAtomic(candidateRow)` is true

done